types/transaction.go

```diff
@@ -29,16 +73,22 @@ func (t *Transaction) IsContractCreation() bool {
 	return t.To == nil
 }
 
+func (t *Transaction) IsLegacyTx() bool {
+	return t.Type == LegacyTx
+}
+
+func (t *Transaction) IsStateTx() bool {
+	return t.Type == StateTx
+}
+
 // ComputeHash computes the hash of the transaction
 func (t *Transaction) ComputeHash() *Transaction {
-	ar := marshalArenaPool.Get()
 	hash := keccak.DefaultKeccakPool.Get()
 
-	v := t.MarshalRLPWith(ar)
-	hash.WriteRlp(t.Hash[:0], v)
-
-	marshalArenaPool.Put(ar)
-	keccak.DefaultKeccakPool.Put(hash)
+	if _, err := hash.Write(t.MarshalRLP()); err == nil {
```

Comment:
What happens if an error occurs on `Write`? Hash will not be computed?

Reply:
I have reverted this function to original one. I have just copy pasted it from original state tx branch but the change does not make any difference. We should probably tackle this problem in the future.